source/portfolio/market_data_mad.h

```diff
@@ -0,0 +1,30 @@
+//
+// Created by eduar on 06/04/2021.
+//
+
+#ifndef PORTFOLIO_MARKET_DATA_MAD_H
+#define PORTFOLIO_MARKET_DATA_MAD_H
+#include "market_data.h"
+#include <ostream>
+namespace portfolio {
+    class market_data_mad : public market_data {
+      public:
+        market_data_mad(const std::vector<std::string> &assetList,
+                        data_feed &df, const minute_point &startPeriod,
+                        const minute_point &endPeriod, timeframe tf,
+                        int mad_periods);
+        friend std::ostream &operator<<(std::ostream &os,
+                                        const market_data_mad &mad);
+        std::vector<std::string>::iterator assets_begin();
+        std::vector<std::string>::iterator assets_end();
+        double risk_of(const std::string &asset);
+        double expected_return_of(const std::string &asset);
+
+      private:
+        std::vector<std::string> assets_;
```

Comment:
Talvez uma opção seja remover o `assets_` totalmente. Note que esta estrutura de dados está replicando 3 cópias de todos os seus assets. Você pode achar se o asset existe no `risk_` ou no `expected_return_`, então pode remover o `assets_`. Já o risk and expected return podem ir para um `std::map<std::string, std::pair<double,double>>` só para evitar replicar as `std::string`s dos assets. Note também que essa string é sua operação mais cara e importante de evitar aqui. Cada string criada é uma alocação nova no heap.

Reply:
Conforme sugestão reorganizei o código para usar o std::map<std::string, std::pair<double,double>>.